### PR TITLE
🐛 Fixes unset/empty write file content in cloudinit

### PR DESCRIPTION
**Pull request by srm09**
**What does this PR do, and why is it needed?**
This patch fixes the validation logic to include the content in the CloudInit WriteFiles section to be empty. Since the API declares it as a required field, we set it to empty during the marshaling process to generate the YAML cloudconfig document.

**Which issue(s) is/are addressed by this PR?**:
n/a

**Are there any special notes for your reviewer**:
n/a

**Please add a release note if necessary**:
```release-note
🐛 Fixes empty or unset write file content in cloudinit
```

### CODE REVIEW

File: pkg/util/cloudinit/cloudconfig_test.go [hunk 1]
```diff
@@ -152,6 +152,79 @@ var _ = Describe("CloudConfig MarshalYAML", func() {
...
+				Expect(err).ToNot(HaveOccurred())
+				Expect(data).To(Equal(cloudConfigWithEmptyFileContent))
+			})
+		})
+
+		When("Write File content is empty", func() {
+			BeforeEach(func() {
+				cloudConfig.WriteFiles = append(cloudConfig.WriteFiles, vmopv1cloudinit.WriteFile{
+					Path:        "/foo",
+					Content:     []byte(``),
```
Comment (akutz) [inline pkg/util/cloudinit/cloudconfig_test.go:213]: Please have a test where the `Content` field is set to `nil`

**Review by akutz — APPROVED** [review]
lgtm, but for one test. Thank you!

File: pkg/util/cloudinit/validate/validate_test.go [hunk 2]
```diff
@@ -97,6 +97,24 @@ var _ = Describe("Validate CloudConfigJSONRawMessage", func() {
...
+			cloudConfig.WriteFiles[0].Content = nil
+		})
+		It("Should not return any errors", func() {
+			Expect(errs).To(HaveLen(0))
+		})
+	})
+
+	When("The CloudConfig's first file content is empty", func() {
+		BeforeEach(func() {
+			cloudConfig.WriteFiles[0].Content = []byte(``)
```
Comment (akutz) [inline pkg/util/cloudinit/validate/validate_test.go:112]: Also please have a test where this field is `nil` and not `[]byte(``)`

**Review by sreyasn — APPROVED** [review]
LGTM. Thanks Sagar

File: pkg/util/cloudinit/validate/validate_test.go (hunk 2 above)
Comment (srm09) [reply pkg/util/cloudinit/validate/validate_test.go:112]: https://github.com/vmware-tanzu/vm-operator/pull/334/files#diff-71149d5c5d525b6bff3a89a726ea2fd1bc114febde84de2b25bb4feb79012ba2R103

Are you thinking of something other than the one added?

Comment (akutz) [reply pkg/util/cloudinit/validate/validate_test.go:112]: Somehow I missed that, thanks!

File: pkg/util/cloudinit/cloudconfig.go [hunk 3]
```diff
@@ -139,9 +139,11 @@ func MarshalYAML(
 			// a string from the Content field.
 			content := secret.WriteFiles[in.WriteFiles[i].Path]
 			if content == "" {
-				if err := json.Unmarshal(
-					in.WriteFiles[i].Content, &content); err != nil {
-
+				inContent := in.WriteFiles[i].Content
+				if len(inContent) == 0 {
```
Comment (bryanv) [inline pkg/util/cloudinit/cloudconfig.go:143]: Not a big deal but I think `if inContent == nil` is a little clearer here (or otherwise won't we assign the same thing below?)

**Review by bryanv — APPROVED** [review]

File: pkg/util/cloudinit/cloudconfig.go (hunk 3 above)
Comment (akutz) [reply pkg/util/cloudinit/cloudconfig.go:143]: No.  `[]byte(`""`)` is not the same as `len(inContent) == 0`

Comment (akutz) [reply pkg/util/cloudinit/cloudconfig.go:143]: In the above case, the `inContent` is literally the JSON as it would be written in a JSON doc. It's "raw JSON" as it were.

Comment (akutz) [reply pkg/util/cloudinit/cloudconfig.go:143]: Therefore even `[]byte{}` would be invalid since it is not an "empty string" per the JSON rules.
